File: CSharp/DependencyBreakingKatas/C/Money.cs
namespace Org.Codecop.Dependencies.C
{
    public class Money
    {
        private readonly decimal value;

        public Money(int value) => this.value = value;

        public Money(decimal value) => this.value = value;

        public Money Add(Money other)
        {
            return new Money(value + other.value);
        }

        public Money Percentage(int p)
        {
            return new Money(value * p / 100);
        }

        public string Format()
        {
            return string.Format("{0:0.00}", value);
        }

        internal decimal AsDecimal()
        {
            return value;
        }
    }
}

File: Java/src/main/java/org/codecop/dependencies/c/Money.java
package org.codecop.dependencies.c;

import java.math.BigDecimal;
import java.util.Objects;

public class Money {

    private final BigDecimal value;

    public Money(int value) {
        this(new BigDecimal(value));
    }

    public Money(String value) {
        this(new BigDecimal(value));
    }
    
    public Money(BigDecimal value) {
        this.value = value;
    }

    public Money add(Money other) {
        return new Money(value.add(other.value));
    }

    public Money percentage(int p) {
        return new Money(value.multiply(new BigDecimal(p)).divide(new BigDecimal(100)));
    }

    public String format() {
        return String.format("%.2f", value);
    }
    
    @Override
    public boolean equals(Object other) {
        if (other == null || getClass() != other.getClass()) {
            return false;
        }
        Money money = (Money) other;
        return Objects.equals(value, money.value);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(value);
    }
    
    public BigDecimal asBigDecimal() {
        return value;
    }
}
